addons/product_images/wizard/product_fetch_image_wizard.py

```diff
@@ -0,0 +1,264 @@
+            lambda p: not p.image_1920 and p.barcode and p.image_fetch_pending
+        )[:limit]  # limit to apply after the filter when self.products_to_process is used.
+
+    def _cron_fetch_image(self):
+        """ Fetch images of a list of products using their barcode
+
+        This function is called from a cron job. If the daily request limit is reach, the cron job
+        will be schedule to run again a day later.
+        """
+        # Retrieve 100 products at a time to limit the cron's time length and avoid reach Google
+        # default rate limit.
+        products_to_process = self._get_products_to_process(100)
+        try:
+            self._fetch_image_from_google(products_to_process)
+        except HTTPError as http_error:
+            if http_error.response.status_code == requests.codes.too_many_requests:
+                self.env.ref('product_images.ir_cron_fetch_image')._trigger(
+                    fields.Datetime.now() + timedelta(days=1.0)
+                )
+            else:
+                raise
+        except ConnectionError:
+            self.env.ref('product_images.ir_cron_fetch_image')._trigger(
+                fields.Datetime.now() + timedelta(hours=1.0)
+            )
+        else:
+            if self._get_products_to_process(1):
+                self.env.ref('product_images.ir_cron_fetch_image')._trigger(
+                    fields.Datetime.now() + timedelta(minutes=1.0)
+                )
+
+    def _fetch_image_from_google(self, products):
+        """ Fetch images of a list of products from Google
+
+        Retrieve an image for each product in `products` from the Google Custom Search API
+        using their barcode. Fetch the 10 first images url and save the first valid image.
+
+        :param recordset products: The products for which an image must be fetched, as a
+        `product.product` recordset
+        :return: None
+        """
+        if not products:
+            return
+
+        ICP = self.env['ir.config_parameter']
+        session = requests.Session()
+        custom_search_req = requests.Request(
+            method='GET',
+            url='https://customsearch.googleapis.com/customsearch/v1',
+            params={
+                'cx': ICP.get_param('google.pse.id').strip(),
+                'safe': 'active',
+                'searchType': 'image',
+                'key': ICP.get_param('google.custom_search.key').strip(),
+                'rights': 'cc_publicdomain,cc_attribute,cc_sharealike',
+                'imgSize': 'large',
+                'imgType': 'photo',
+                'fields': 'searchInformation/totalResults,items(link)',
+            },
+        )
+        prepared_req = session.prepare_request(custom_search_req)
+        request_url = prepared_req.url
+        for product in products:
+            try:
+                prepared_req.prepare_url(request_url, {'q': product.barcode})
+
+                response = session.send(prepared_req, timeout=5)
```

Comment:
This entire thing seems weird. Why not just `session.get()` with a dict of params? The possible optimisation of only creating one request object seems... unnecessary.